src/Provider/FreeGeoIp/FreeGeoIp.php

```diff
@@ -83,4 +94,36 @@ public function getName(): string
     {
         return 'free_geo_ip';
     }
+
+    /**
+     * @param string $locale
+     */
+    public function setLocale(string $locale)
```

Comment:
This makes the provider stateful. That is something we should avoid in 4.0.  I suggest adding this as constructor parameter. 

Also if the Query object has a locale parameter we should override the `$this->locale`